### DISCUSSION TITLE
[NEW FEATURE]Creation of Specific Schemas for Each Team

### DISCUSSION THREAD

**Discussion by lduf** [org-SCAN/website#514] [Ressource]
## **Description:**
This feature introduces the ability to dynamically create specific database schemas for individual teams. Each team will have its own schema, including custom tables such as `fields`, `items`, and `links`. The process will rely on Artisan commands to manage schema creation and migration execution, ensuring flexibility and modularity in the system. Tests will validate all functionalities to ensure a seamless experience.

---

## **Tasks to Perform:**

1. **Artisan Command to Generate Specific Schemas:**
   - Implement an Artisan command:
     ```bash
     php artisan make:schema TEAM_ID
     ```
   - This command must:
     - Verify if a schema for `TEAM_ID` already exists.
     - Create a new schema in the database if it does not exist.

2. **Organize Schema-Specific Migrations:**
   - Establish a directory structure for migrations:
     - **`/database/migrations/global`**: Contains global migrations.
     - **`/database/migrations/teams/TEAM_ID`**: Contains migrations specific to a team.

3. **Artisan Command to Copy and Execute Base Table Migrations:**
   - Implement an Artisan command:
     ```bash
     php artisan migration:copy TABLE_NAME SCHEMA
     ```
   - This command must:
     - Copy all required global migration files into the directory `teams/TEAM_ID`.
     - Execute the copied migrations in the specified schema.

4. **Handle Global Migration Files:**
   - **Approach: Direct Copy of Global Files**
     - Identify all global migrations associated with a table and copy them into the team-specific schema directory.
     - Retain the complete history of table changes.

5. **Schema Migration Execution:**
   - Execute all migrations in a specified schema using an Artisan command.

6. **Unit and Functional Testing:**
   - Develop tests to validate the following functionalities:
     - Creating a specific schema.
     - Executing migrations within a specific schema.
     - Verifying the creation of the necessary tables.
   - Test cases:
     - A team with an existing schema.
     - A team requiring multiple global migration files for a single table.

[a comment by an automated account is not shown]

**Comment by vqlion** [comment]
### Some comments on the tasks to perform:

> 1. **Artisan Command to Generate Specific Schemas:**
>    - Implement an Artisan command:
>      ```bash
>      php artisan make:schema TEAM_ID
>      ```
>    - This command must:
>      - Verify if a schema for `TEAM_ID` already exists.
>      - Create a new schema in the database if it does not exist.

So to be clear, "create a new schema" means creating a new database in the MySQL instance and creating all the table we need? I'm not sure what "schema" refers to in this context.

> 3. **Artisan Command to Copy and Execute Base Table Migrations:**
>    - Implement an Artisan command:
>      ```bash
>      php artisan migration:copy TABLE_NAME SCHEMA
>      ```
>    - This command must:
>      - Copy all required global migration files into the directory `teams/TEAM_ID`.
>      - Execute the copied migrations in the specified schema.
> 
> 4. **Handle Global Migration Files:**
>    - **Approach: Direct Copy of Global Files**
>      - Identify all global migrations associated with a table and copy them into the team-specific schema directory.
>      - Retain the complete history of table changes.

I understand what this command does, but I don't understand why we would need it. I think it's nice to have a global migrations folder and team-specific ones, but I don't see why we would need to copy migrations files from global to specific. 
To me "global migrations" describe migrations on the global scope, so in the global dB, and not the team-specific ones, and "team-specific migrations" describe migrations on the specific scope. 
Maybe on team creation it could be useful to copy lists or pre-defined item-types?

### The process I thought of:

- On team creation:
   - Create the new dB for the team
   - Create the migration files and execute the migrations for this team
   => Result is a dB for the new team with all the tables ready and empty (`item_types`, `fields`, `duplicates`, `relation`,...)

- Then init the team:
   - Run a command to create the tables we need, like tables for items or lists defined in the global scope
   - The code for this will exist anyway, because we need a way to go from the `item_types` table to having one table for each item type
   => Result is the same dB as before with the tables from the global scope dynamically created

- On new table creation (e.g. creating a new item_type)
   - Run the same code as above, designed to simply ignore if the table already exists with the right fields
   - If a new item type is created in the global scope again it's the same principle

With this I don't think we need migration files for each team, but some sort of TableController that does the heavy-lifting and handles edge cases. It should be transparent from a specific-scope-pov whether an item-type comes from the global or the specific scope, so the controller would handle that.

Let me know what you think :)

**Comment by lduf** [comment]
> So to be clear, “create a new schema” means creating a new database in the MySQL instance and creating all the tables we need? I’m not sure what “schema” refers to in this context.

I’m using the definition of schema as outlined in the MySQL Glossary:
https://dev.mysql.com/doc/refman/8.4/en/glossary.html#glos_schema

I*n MySQL, physically, a schema is synonymous with a database. You can substitute the keyword SCHEMA for DATABASE in MySQL SQL syntax, for example by using CREATE SCHEMA instead of CREATE DATABASE.*

What I mean here, however, is creating the global structure, without the tables.
I would like artisan commands that focus on very specific tasks.
We start by creating the schema (database), and then proceed to create the tables.

---

> I understand what this command does, but I don’t understand why we would need it. I think it’s nice to have a global migrations folder and team-specific ones, but I don’t see why we would need to copy migration files from global to specific.
To me, “global migrations” describe migrations on the global scope, so in the global database, and not the team-specific ones. “Team-specific migrations” describe migrations on a more specific scope.
Maybe it could be useful to copy lists or predefined item types when creating a new team?

Global migrations define tables that we need to have in the team-specific schema.
For example: fields. Fields will be defined globally through migration files. We might have multiple migration files related to fields because the table structure may have been modified over time.

Copying all the migration files into the specific team’s migration folder ensures that the exact same structure as the global one is replicated. If the table structure differs, inheritance and merging data between the global and team-specific schemas won’t be feasible. It’s essential to guarantee the same structure across schemas.

Since migrations are Laravel’s way of managing database and table structures, this is why I chose to rely on them for this purpose.

---

Regarding your process, I completely agree—it aligns more or less with what I had in mind.

**Comment by vqlion** [comment]
That makes more sense thanks!

So when copying global migrations to specific ones, we should make sure to only copy what we need. Only the migrations for inherited tables must be copied, but migrations for e.g. users or teams tables are not needed on specific scopes (or we end up having empty tables).

Also we'll have tables in specific scope that don't exist in global (like duplicates). So we could have these migrations ready to be copied to specific migration folders as well, but not from the global migrations (since the table doesn't exist in global).

Another thing: is it useful to have team-specific migrations, or would a "specific-scope" migration folder be enough? If we also use migrations to dynamically create the item tables, it's useful to have team-specific folders, but otherwise all the teams should have the same migrations.

**Comment by lduf** [comment]
I think we should have a separate migration folder for each specific schema, structured like this:

```
migrations/
   global/
   {team_id1}/
   {team_id2}/
   ...
```

This structure is essential because when creating a table for a specific team, a new migration will be required. For instance, if you want to introduce a new `Item_type` (Guns) you'll need to create a model and its corresponding migration for the pertinent team.

When it comes to tables meant for a specific scope, I don't like the idea of having empty tables globally only to duplicate their structure locally. Why not having a template directory, which could serve as a more efficient solution as I won't be migrated. Worth to think about it because, having a template directory could imply having to check both in global directory and template directory when copying migrations :cry:

Any idea ?

**Comment by lduf** [comment]
## Vocabulary:

**subscribed to a table**: being part of the schemas that will receive all migration files associated with a given table.
**publish a table**: copy the migrations associated with this table into the migration folder of all subscribers.

## General Idea

A `template` folder will contain the migrations of tables that need to be specific and/or global. We will need a model to allow schemas to subscribe to tables. Thus, when a table is updated, we will have a command that allows the publication of the associated migrations.

## What I thought about

To do this, we need to:

Modify the `migrations` table (automatically generated and managed by Laravel) to include the schemas in which tables are published:

- ~~either directly modify the `migrations` table~~

- or don't touch the `migrations` table and create an adjacent table.

I prefer to have an adjacent table.

As I recall (to verify), the `migrations` table contains the migration filename and a status.

From the migration file, we look at the associated table and store them in a `tables` table (or another name to avoid future confusion). The `tables` table contains the active tables of the project.

> Does `tables` contain all the tables, regardless of the schema?
>
> > intuitively: Yes, but can be discussed

**`tables`**

```
#, name
1, item_type
2, duplicates
3, users
```

We also need a `table_scopes` table to manage schema subscriptions. This table allows associating all the tables and schemas in which we must publish migrations.

**`table_scopes`**

```
#, table_id, schema
1, 1, global
2, 1, team_1
3, 2, team_1
4, 3, global
```

Here item_type is published to global and team_1 schemas while duplicates (resp. users) is only published to team_1 (resp. global).

### Process

When we do a new migration, we have the following process:

- Laravel automatically: adds to the migration table ⚠️ check the behavior with multiple schemas
- Us: we retrieve the associated table and check if it's a new table or not. If yes, we add to `tables`
- Managing scopes: allow a schema to subscribe to a table: add to the `table_scopes` table. First I think we can do it through Models, then we may want an UI to allow user to do it.

Then each new migration is published respectively to its subscribers.

By doing this, we can be precise even beyond the scope: we can specify the schemas in which the tables should be duplicated. This will allow us to publish a migration very precisely. Example: a team wants to override a list e.g., *Roles*, so we subscribe this team to the *Roles* table.

It remains to define if we override the migration mechanism in Laravel:

- Can we directly override the functioning of Laravel -> needs to publishing the migration files and artisan commands from vendor.

Everything remains transparent, and we can continue to use the classic commands, but it may be complicated. I'm not even sure if we can make such modifications. We may struggle with new Laravel updates.

- We create our own files and migration methods

This will force us to no longer use Laravel's classic commands. A very precise doc is needed to explain all this. We also risk increasing the number of programming errors. 
See if worst case, we can redirect commands. Like when we call Laravel migration commands, we do a redirection to our commands.

**Reply by vqlion** [reply]
This works, but it looks very complex either way. Overriding how Laravel works is going to be a pain when we have to update it, and creating our own migration methods feels like rewriting Laravel ^^'

I think we can find a way to make this work with Laravel's commands right away. 

I was thinking that the migration files could define themselves which scope they go to (global/specific/both), that way when Laravel executes the migrations it executes each of them for each relevant schema. We'll need something to manage multi-schema connections, it seems to be well-documented.
This means all the teams will have the same base schema, and only the dynamic tables will change (like item types). I don't think it's a bad thing, it keeps things easy to understand and to manage and I'm not sure there's much use to subscribing to some tables but not others depending on the team you're into. Any _overridable_ list could just be copied on the team's dB.

It would also eliminate the need for multiple migration folders, at least for the base migrations, since a single migration file could handle multiple schemas at once. I don't know if we'd still need the multi-folder structure for dynamic tables, but maybe we would.

Overall I think we should build this around Laravel's features as much as possible. What I suggest is a bit more difficult when adding a migration (you have to make sure you specify the right scope of the migration in the file) but at least Laravel does all the hard work for us after that.

**Comment by lduf** [comment]
Btw, what we are discussing here is out of the scope of this issue. I'm gonna convert this issue to discussion and rescope it once we agree on the architecture.
